### DISCUSSION TITLE
Going from an AST to a PEP8-formatted Python file

### DISCUSSION THREAD

**Discussion by bogdanstate** [RustPython/RustPython#2384] [Q&A]
This is a bit of an unusual use case, but I am trying to generate some human-readable Python code from Rust. I am interested just in having syntactically valid code for the time being, so compilation is not a concern.

Is there any project currently that allows one to format a rustpython AST to a PEP8-compliant file? I have done some digging around the code base here, but I can't seem to find anything to answer this use case. Any pointers much appreciated (as would be confirmation that this functionality is not supported)!

**Comment by bogdanstate** [comment]
Thought about this some more, and I guess writing a "reverse parser" for the AST is rather trivial. That + something like `autopep8` should do the trick, although it feels a bit hacky to be calling a CLI tool.

**Reply by zicklag** [reply]
If you wanted to avoid the CLI tool you might be able to use RustPython or PyO3 to embed the Python interpreter and install `autopep8` embedded in your rust application.

Granted depending on what your use-case is it might not be worth the effort of figuring out how to embed `autopep8`. I saw a guide on how to do stuff like that once, I think, with `pyo3-pack`, but I think its been renamed: https://github.com/PyO3/maturin. But that looks like Rust libraries as Python packages, not the other way around, so I don't know where the instructions I'm thinking of went.

**Comment by charliermarsh** [comment]
If you were in the business of implementing this yourself, there's logic you can follow from astor: https://github.com/berkerpeksag/astor/blob/master/astor/code_gen.py
